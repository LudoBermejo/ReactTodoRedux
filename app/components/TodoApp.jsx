import uuid from 'node-uuid';
import React from 'react';
import moment from 'moment';

import TodoList from 'TodoList';
import AddTodoItem from 'AddTodoItem';
import TodosFilter from 'TodosFilter';
import TodoAPI from 'TodoAPI';

export default class TodoApp extends React.Component {
  constructor(props) {
    super(props);

    this.addTodoItem = this.addTodoItem.bind(this);
    this.search = this.search.bind(this);
    this.changeCompleteTodo = this.changeCompleteTodo.bind(this);
    this.state = {
      searchText: '',
      filterByComplete: false,
      todoList: TodoAPI.getTodos()
    };
  }

  componentDidUpdate() {
    TodoAPI.setTodos(this.state.todoList);
  }

  search(value) {
    this.setState({ ...value });
  }

  addTodoItem(value) {
    const todoList = this.state.todoList;
    this.setState({
      todoList: [
        ...todoList,
        {
          id: uuid(),
          value,
          completed: false,
          createdAt: moment().unix()
        }
      ]
    });
  }

  changeCompleteTodo(id) {
    this.setState({
      todoList: this.state.todoList.map((item) => {
        const todo = item;
        if (todo.id === id) {
          todo.completed = !item.completed;
          todo.completedAt = todo.completed ? moment().unix() : undefined;
        }
        return todo;
      })
    });
  }

  render() {
    const { todoList, searchText, filterByComplete } = this.state;
    const filteredTodo = TodoAPI.searchTodos(todoList, filterByComplete, searchText);

    return (
      <div>
        <h1 className="page-title">Todo App</h1>
        <div className="row">
          <div className="column small-centered small-11 medium-6 large-5">
            <div className="container">
              <TodosFilter onSearch={this.search} />
              <TodoList todoList={filteredTodo} onChangeItemComplete={this.changeCompleteTodo} />
              <AddTodoItem onAddItem={this.addTodoItem} />
            </div>
          </div>
        </div>
      </div>
    );
  }
}

