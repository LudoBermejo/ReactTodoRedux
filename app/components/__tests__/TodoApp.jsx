import uuid from 'node-uuid';
import React from 'react';
import renderer from 'react-test-renderer';
import moment from 'moment';
import { shallow, mount } from 'enzyme';
import TodoApp from 'TodoApp';


describe('Todo app', () => {
  it('renders properly', () => {
    const todoApp = renderer.create(<TodoApp />).toJSON();
    expect(todoApp).toMatchSnapshot();
  });

  it('should add to do item', () => {
    const todoApp = shallow(<TodoApp />);
    const todoList = todoApp.instance().state.todoList;
    const numberOfItems = todoList.length;
    todoApp.instance().addTodoItem('demo');
    expect(todoApp.instance().state.todoList.length).toBe(numberOfItems + 1);
    expect(todoApp.instance().state.todoList[todoApp.instance().state.todoList.length-1].value).toBe('demo');
  });

  it('should change completed item and add specific date', () => {
    const todoApp = shallow(<TodoApp />);
    todoApp.instance().setState({
      todoList: [
        {
          id: uuid(),
          value: 'first',
          createdAt: moment().unix(),
          completed: false,
        },
        {
          id: uuid(),
          value: 'Second',
          createdAt: moment().unix(),
          completed: false
        },
        {
          id: uuid(),
          value: 'Third',
          createdAt: moment().unix(),
          completed: false
        },
        {
          id: uuid(),
          value: 'Fourth',
          createdAt: moment().unix(),
          completed: false
        }
      ]
    });

    expect(todoApp.instance().state.todoList[0].completed).toBeFalsy();
    expect(todoApp.instance().state.todoList[0].completedAt).toBeUndefined();
    todoApp.instance().changeCompleteTodo(todoApp.instance().state.todoList[0].id);
    expect(todoApp.instance().state.todoList[0].completed).toBeTruthy();
    expect(todoApp.instance().state.todoList[0].completedAt).toBeDefined();

  });

  it('should remove completedAt if item is not completed', () => {
    const todoApp = shallow(<TodoApp />);
    todoApp.instance().setState({
      todoList: [
        {
          id: uuid(),
          value: 'first',
          createdAt: moment().unix(),
          completed: true,
          completedAt: moment().unix(),
        }
      ]
    });

    expect(todoApp.instance().state.todoList[0].completedAt).toBeDefined();
    todoApp.instance().changeCompleteTodo(todoApp.instance().state.todoList[0].id);
    expect(todoApp.instance().state.todoList[0].completed).toBeFalsy();
    expect(todoApp.instance().state.todoList[0].completedAt).toBeUndefined();

  });

  it('should store data even if you refresh', () => {
    const todoApp = shallow(<TodoApp />);
    todoApp.instance().setState({
      todoList: [
      ]
    });
    expect(todoApp.instance().state.todoList.length).toBe(0);

    todoApp.instance().setState({
      todoList: [
        {
          id: uuid(),
          value: 'Fourth',
          createdAt: moment().unix(),
          completed: false
        }
      ]
    });
    expect(todoApp.instance().state.todoList.length).toBe(1);

    const todoApp2 = shallow(<TodoApp />);
    expect(todoApp2.instance().state.todoList.length).toBe(1);

  });



});

