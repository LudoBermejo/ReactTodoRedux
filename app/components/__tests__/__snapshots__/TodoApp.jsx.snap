exports[`Todo app renders properly 1`] = `
<div>
  <h1>
    Todo
  </h1>
  <div>
    <div>
      <input
        onChange={[Function]}
        placeholder="Search in todos"
        type="search" />
    </div>
    <label>
      <input
        onChange={[Function]}
        type="checkbox" />
    </label>
  </div>
  <div>
    <h1>
      List of items:
    </h1>
    <ul />
  </div>
  <form
    className="form"
    onSubmit={[Function]}>
    <input
      placeholder="What do you need to do?"
      type="text" />
    <button
      className="button expanded">
      Add todo
    </button>
  </form>
</div>
`;
